# Data Manipulation with Pandas  
1. Transforming DataFrames
2. Aggregating DataFrames
3. Slicing and Indexing DataFrames
4. Creating and Visualizing DataFrames

### 1. Transforming DataFrames  
  
**What´s the point of pandas?**  
* Data Manipulation
 * Ch 1, Sorting and Subsetting, Creating new columns
 * Ch 2, Summary statistics, Counting, Grouped Summary statistics
 * Ch 3, Subsetting using slicing, Indexing and subsetting using index
* Data Visualization  
 * Ch 4, Plotting, Handling missing data, Reading data into a DataFrame
  
  
* **Introducing DataFrames**  
pandas is built on top of NumPy and Matplotlib.  
  
Rectangular data or tabular data is the a common form of data pandas is design to work with it.  
Each row is an observation and each column is a property or attribute. In pandas this rectangular data is represented as a `DataFrame` object. Other programming languages has something similar, `SQL` has `data table` and `R` also has `data.frame`. Every column have the same data type on each row but different columns can have different data types.  
  
Exploring a DataFrame object using methods/attributes:  
  
`df.head()` method displays the first rows (default is 5)  
`df.indo()` method displays the names of columns, data types they contain and if there are any missing alues.  
`df.shape` attribute shows the number of rows followed by the number of clolumns, e.g `(8,7)`  
`df.describe()` method calculates some summary statistics for numerical columns.  
  
A `DataFrame` consists of three different components accessible using attributes.  
  
`df.values` attribute contains the data values in a 2D numpy array  
`df.columns` attribute contains the labels of the column names as index object.  
`df.index` attribute contains row numbers of row names as index object. Note: Row labes are stored in `.index` and not in `.rows`!  
  
**pandas Philosophy**  
There should be one -- and preferably only one -- obvious way to do it!  
*- The Zen of Python* by Tim Peters, Item 13 https://www.python.org/dev/peps/pep-0020/ 
Pandas deliberately does not follow this philosphoy, instead there are often many ways to solve a problem in pandas and we can choose the best one. pandas gives many tools to give options.

**Inspecting a DataFrame**  
When we have a new DataFrame to work with, the first thing you need to do is explore it and see what it contains. There are several useful methods and attributes for this.  
  
`.head()` returns the first few rows (the “head” of the `DataFrame`).  
`.info()` shows information on each of the columns, such as the data type and number of missing values.  
`.shape` returns the number of rows and columns of the DataFrame.  
`.describe()` calculates a few summary statistics for each column.

In [11]:
import pandas as pd

# OMA Need to create the DF
path = "data/homelessness.csv"
homelessness = pd.read_csv(path, index_col=0)

In [12]:
# Print the head of the homelessness data
print(homelessness.head())

               region       state  individuals  family_members  state_pop
0  East South Central     Alabama       2570.0           864.0    4887681
1             Pacific      Alaska       1434.0           582.0     735139
2            Mountain     Arizona       7259.0          2606.0    7158024
3  West South Central    Arkansas       2280.0           432.0    3009733
4             Pacific  California     109008.0         20964.0   39461588


In [3]:
# Print information about homelessness
print(homelessness.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51 entries, 0 to 50
Data columns (total 6 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Unnamed: 0      51 non-null     int64  
 1   region          51 non-null     object 
 2   state           51 non-null     object 
 3   individuals     51 non-null     float64
 4   family_members  51 non-null     float64
 5   state_pop       51 non-null     int64  
dtypes: float64(2), int64(2), object(2)
memory usage: 2.5+ KB
None


In [4]:
# Print the shape of homelessness
print(homelessness.shape)

(51, 6)


In [6]:
# Print a description of homelessness
print(homelessness.describe())

       Unnamed: 0    individuals  family_members     state_pop
count   51.000000      51.000000       51.000000  5.100000e+01
mean    25.000000    7225.784314     3504.882353  6.405637e+06
std     14.866069   15991.025083     7805.411811  7.327258e+06
min      0.000000     434.000000       75.000000  5.776010e+05
25%     12.500000    1446.500000      592.000000  1.777414e+06
50%     25.000000    3082.000000     1482.000000  4.461153e+06
75%     37.500000    6781.500000     3196.000000  7.340946e+06
max     50.000000  109008.000000    52070.000000  3.946159e+07


We see the average number of homeless individuals of each state is about 7226.  
Now we do some more exploration of the df!  
  
**Parts of a DataFrame**  
To better understand DataFrame objects, it's useful to know that they consist of three components, stored as attributes:  
  
* `.values`: A two-dimensional NumPy array of values.  
* `.columns`: An index of columns: the column names.  
* `.index`: An index for the rows: either row numbers or row names.  
  
You can usually think of indexes as a list of strings or numbers, though the pandas `Index` data type allows for more sophisticated options. (These will be covered later in the course.)  
  
* Print a 2D NumPy array of the values in `homelessness`
* Print the column names of `homelessness`
* Print the index of `homelessness`

In [13]:
# Print the values of homelessness
print(homelessness.values)

[['East South Central' 'Alabama' 2570.0 864.0 4887681]
 ['Pacific' 'Alaska' 1434.0 582.0 735139]
 ['Mountain' 'Arizona' 7259.0 2606.0 7158024]
 ['West South Central' 'Arkansas' 2280.0 432.0 3009733]
 ['Pacific' 'California' 109008.0 20964.0 39461588]
 ['Mountain' 'Colorado' 7607.0 3250.0 5691287]
 ['New England' 'Connecticut' 2280.0 1696.0 3571520]
 ['South Atlantic' 'Delaware' 708.0 374.0 965479]
 ['South Atlantic' 'District of Columbia' 3770.0 3134.0 701547]
 ['South Atlantic' 'Florida' 21443.0 9587.0 21244317]
 ['South Atlantic' 'Georgia' 6943.0 2556.0 10511131]
 ['Pacific' 'Hawaii' 4131.0 2399.0 1420593]
 ['Mountain' 'Idaho' 1297.0 715.0 1750536]
 ['East North Central' 'Illinois' 6752.0 3891.0 12723071]
 ['East North Central' 'Indiana' 3776.0 1482.0 6695497]
 ['West North Central' 'Iowa' 1711.0 1038.0 3148618]
 ['West North Central' 'Kansas' 1443.0 773.0 2911359]
 ['East South Central' 'Kentucky' 2735.0 953.0 4461153]
 ['West South Central' 'Louisiana' 2540.0 519.0 4659690]
 ['New 

In [14]:
# Print the column index of homelessness
print(homelessness.columns)

Index(['region', 'state', 'individuals', 'family_members', 'state_pop'], dtype='object')


In [15]:
# Print the row index of homelessness
print(homelessness.index)

Int64Index([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
            17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33,
            34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49,
            50],
           dtype='int64')


A `DataFrame` has three components: values, a column index, and a row index  
  
**Sorting and subsetting**  
  
Sorting - change the order of the row by sorting them.  
  
`df.sort_values("col_name")` a method that takes column_name as argument  
`df.sort_values(["col_name_1", "col_name_2"], ascending = [True, False])` arguments for multiple columns and sorting order
`df["col_name"]`, to look at only one column
`df[["col_name_1"],["col_name_2"]]`, the outer squared brackets are responsible for subsetting the data frame, the inner squared brackets are responsible for creating a list of column names to subset.  
`cols_to_subset = ["col_name_1", "col_name_2"]` , create the list of columns as a variable   
`df[cols_to_subset]`, use the separate list as a variable to subset the df  
  
Subsetting rows - by creating logic to filter by  
`df[df["col_name_1"] > 50]` , subset the df with a logic of a column. We can sub-set also with text and dates  
  
Subsetting using multiple conditions using operators:  
`is_yes = df["col_name_1"] == "yes"`  
`is_no = df["col_name_2"] == "no"`  
`df[is_yes & is_no]`  
  
Or as one line of code, need paranthesis around each condition:  
`df[ (df["col_name_1"] == "yes") & (df["col_name_1"] == "no") ]`  
  
Subsetting using categorical values using `isin()` method:  
`is_black_or_brown = df["color"].isin(["Black", "Brown"])`  
`df[is_black_or_brown]`  

**Sorting rows**  
Examples

In [16]:
homelessness.head()

region       state  individuals  family_members  state_pop
0  East South Central     Alabama       2570.0           864.0    4887681
1             Pacific      Alaska       1434.0           582.0     735139
2            Mountain     Arizona       7259.0          2606.0    7158024
3  West South Central    Arkansas       2280.0           432.0    3009733
4             Pacific  California     109008.0         20964.0   39461588

Sort `homelessness` by the number of homeless `individuals` in ascending order, and save this as `homelessness_ind`.

In [17]:
# Sort homelessness by individual
homelessness_ind = homelessness.sort_values(["individuals"])

# Print the top few rows
print(homelessness_ind.head())

                region         state  individuals  family_members  state_pop
50            Mountain       Wyoming        434.0           205.0     577601
34  West North Central  North Dakota        467.0            75.0     758080
7       South Atlantic      Delaware        708.0           374.0     965479
39         New England  Rhode Island        747.0           354.0    1058287
45         New England       Vermont        780.0           511.0     624358


Sort `homelessness` by the number of homeless `family_members` in descending order, and save this as `homelessness_fam`.

In [18]:
# Sort homelessness by descending family members
homelessness_fam = homelessness.sort_values(["family_members"], ascending = False)

# Print the top few rows
print(homelessness_fam.head())

                region          state  individuals  family_members  state_pop
32        Mid-Atlantic       New York      39827.0         52070.0   19530351
4              Pacific     California     109008.0         20964.0   39461588
21         New England  Massachusetts       6811.0         13257.0    6882635
9       South Atlantic        Florida      21443.0          9587.0   21244317
43  West South Central          Texas      19199.0          6111.0   28628666


Sort `homelessness` first by region (ascending), and then by number of family members (descending). Save this as `homelessness_reg_fam`.

In [19]:
# Sort homelessness by region, then descending family members
homelessness_reg_fam = homelessness.sort_values(["region", "family_members"], ascending = [True, False])

# Print the top few rows
print(homelessness_reg_fam.head())

                region      state  individuals  family_members  state_pop
13  East North Central   Illinois       6752.0          3891.0   12723071
35  East North Central       Ohio       6929.0          3320.0   11676341
22  East North Central   Michigan       5209.0          3142.0    9984072
49  East North Central  Wisconsin       2740.0          2167.0    5807406
14  East North Central    Indiana       3776.0          1482.0    6695497


Using `.sort_values()` and `.head()` is a powerful pair.

**Subsetting columns**  
When working with data, you may not need all of the variables in your dataset. Square brackets (`[]`) can be used to select only the columns that matter to you in an order that makes sense to you. To select only `"col_a"` of the DataFrame `df`, use  
`df["col_a"]`  
To select two columns use  
`df[["col_a","col_b"]]`  

In [20]:
# Select the individuals column
individuals = homelessness["individuals"]

# Print the head of the result
print(individuals.head())

0      2570.0
1      1434.0
2      7259.0
3      2280.0
4    109008.0
Name: individuals, dtype: float64


In [21]:
# Select the state and family_members columns
state_fam = homelessness[["state","family_members"]]

# Print the head of the result
print(state_fam.head())

        state  family_members
0     Alabama           864.0
1      Alaska           582.0
2     Arizona          2606.0
3    Arkansas           432.0
4  California         20964.0


In [22]:
# Select only the individuals and state columns, in that order
ind_state = homelessness[["individuals","state"]]

# Print the head of the result
print(ind_state.head())

   individuals       state
0       2570.0     Alabama
1       1434.0      Alaska
2       7259.0     Arizona
3       2280.0    Arkansas
4     109008.0  California


**Subsetting rows**  
A large part of data science is about finding which bits of your dataset are interesting. One of the simplest techniques for this is to find a subset of rows that match some criteria. This is sometimes known as * filtering rows*  or *selecting rows*.  
  
There are many ways to subset a DataFrame, perhaps the most common is to use relational operators to return `True` or `False` for each row, then pass that inside square brackets.  
  
`df[df["height_cm"] > 60]`  
`df[df["color"] == "tan"]`  
  
We can filter for multiple conditions at once by using the "bitwise and" operator, `&`.  
  
`df[(df["height_cm"] > 60) & (df["color"] == "tan")]`

In [23]:
# Filter for rows where individuals is greater than 10000
ind_gt_10k = homelessness[homelessness["individuals"] > 10000]

# See the result
print(ind_gt_10k)

                region       state  individuals  family_members  state_pop
4              Pacific  California     109008.0         20964.0   39461588
9       South Atlantic     Florida      21443.0          9587.0   21244317
32        Mid-Atlantic    New York      39827.0         52070.0   19530351
37             Pacific      Oregon      11139.0          3337.0    4181886
43  West South Central       Texas      19199.0          6111.0   28628666
47             Pacific  Washington      16424.0          5880.0    7523869


In [24]:
# Filter for rows where region is Mountain
mountain_reg = homelessness[homelessness["region"] == "Mountain"]

# See the result
print(mountain_reg)

      region       state  individuals  family_members  state_pop
2   Mountain     Arizona       7259.0          2606.0    7158024
5   Mountain    Colorado       7607.0          3250.0    5691287
12  Mountain       Idaho       1297.0           715.0    1750536
26  Mountain     Montana        983.0           422.0    1060665
28  Mountain      Nevada       7058.0           486.0    3027341
31  Mountain  New Mexico       1949.0           602.0    2092741
44  Mountain        Utah       1904.0           972.0    3153550
50  Mountain     Wyoming        434.0           205.0     577601


In [25]:
# Filter for rows where family_members is less than 1000 
# and region is Pacific
fam_lt_1k_pac = homelessness[ (homelessness["family_members"] < 1000) & (homelessness["region"]=="Pacific") ]

# See the result
print(fam_lt_1k_pac)

    region   state  individuals  family_members  state_pop
1  Pacific  Alaska       1434.0           582.0     735139


Using square brackets plus logical conditions is often the most powerful way of identifying interesting rows of data.

**Subsetting rows by categorical variables**  
Subsetting data based on a categorical variable often involves using the "or" operator (`|`) to select rows from multiple categories. This can get tedious when you want all states in one of three different regions, for example. Instead, use the `.isin()` method, which will allow you to tackle this problem by writing one condition instead of three separate ones.  
  
`colors = ["brown", "black", "tan"]`  
`condition = df["color"].isin(colors)`  
`df[condition]`

In [26]:
homelessness.head()

region       state  individuals  family_members  state_pop
0  East South Central     Alabama       2570.0           864.0    4887681
1             Pacific      Alaska       1434.0           582.0     735139
2            Mountain     Arizona       7259.0          2606.0    7158024
3  West South Central    Arkansas       2280.0           432.0    3009733
4             Pacific  California     109008.0         20964.0   39461588

In [27]:
# Subset for rows in South Atlantic or Mid-Atlantic regions
south_mid_atlantic = homelessness[homelessness["region"].isin(["South Atlantic","Mid-Atlantic"])]

# See the result
print(south_mid_atlantic)

            region                 state  individuals  family_members  \
7   South Atlantic              Delaware        708.0           374.0   
8   South Atlantic  District of Columbia       3770.0          3134.0   
9   South Atlantic               Florida      21443.0          9587.0   
10  South Atlantic               Georgia       6943.0          2556.0   
20  South Atlantic              Maryland       4914.0          2230.0   
30    Mid-Atlantic            New Jersey       6048.0          3350.0   
32    Mid-Atlantic              New York      39827.0         52070.0   
33  South Atlantic        North Carolina       6451.0          2817.0   
38    Mid-Atlantic          Pennsylvania       8163.0          5349.0   
40  South Atlantic        South Carolina       3082.0           851.0   
46  South Atlantic              Virginia       3928.0          2047.0   
48  South Atlantic         West Virginia       1021.0           222.0   

    state_pop  
7      965479  
8      701547  
9 

In [28]:
# The Mojave Desert states
canu = ["California", "Arizona", "Nevada", "Utah"]

# Filter for rows in the Mojave Desert states
mojave_homelessness = homelessness[homelessness["state"].isin(canu)]

# See the result
print(mojave_homelessness)

      region       state  individuals  family_members  state_pop
2   Mountain     Arizona       7259.0          2606.0    7158024
4    Pacific  California     109008.0         20964.0   39461588
28  Mountain      Nevada       7058.0           486.0    3027341
44  Mountain        Utah       1904.0           972.0    3153550


Using `.isin()` makes subsetting categorical variables a breeze  
  
**New Columns**  
  
Often we want to add new columns derived from existing columns. Adding a new column can go by many names like mutating a data frame, transforming a data frame or feature engineering.  
  
Example of how to add a new column from existing columns:  
`df["heigh_m"] = df["height_cm"]  / 100`  
On the left hand side we have in square brackets the name of the column we want to create.  
On the right hand side we have the calculation.  
We will then have both the existing and the new column in the data frame.  
  
**Multiple manipulations**  
Skinny dogs e.g  
`bmi_lt_100 = df[df["bmi] < 100]`  
`bmi_lt_100_height = bmi_lt_100.sort_values("height_cm", ascending=False)`  
`bmi_lt_100_height[["name", "height_cm", "bmi"]]`  
  
Some examples:

**Adding new columns**  
You aren't stuck with just the data you are given. Instead, you can add new columns to a DataFrame. This has many names, such as *transforming*, *mutating*, and *feature engineering*.  
  
You can create new columns from scratch, but it is also common to derive them from other columns, for example, by adding columns together or by changing their units.

In [29]:
# Add total col as sum of individuals and family_members
homelessness["total"] = homelessness["individuals"] + homelessness["family_members"]

# Add p_individuals col as proportion of individuals
homelessness["p_individuals"] = homelessness["individuals"] / homelessness["total"]

# See the result
print(homelessness.head())

               region       state  individuals  family_members  state_pop  \
0  East South Central     Alabama       2570.0           864.0    4887681   
1             Pacific      Alaska       1434.0           582.0     735139   
2            Mountain     Arizona       7259.0          2606.0    7158024   
3  West South Central    Arkansas       2280.0           432.0    3009733   
4             Pacific  California     109008.0         20964.0   39461588   

      total  p_individuals  
0    3434.0       0.748398  
1    2016.0       0.711310  
2    9865.0       0.735834  
3    2712.0       0.840708  
4  129972.0       0.838704  


If our dataset doesn't have the exact columns you need, we can often make your own from what you have.  
  
**Combo-attack!**  
You've seen the four most common types of data manipulation: sorting rows, subsetting columns, subsetting rows, and adding new columns. In a real-life data analysis, you can mix and match these four manipulations to answer a multitude of questions.  
  
In this exercise, you'll answer the question, "Which state has the highest number of homeless individuals per 10,000 people in the state?"  

In [30]:
# Create indiv_per_10k col as homeless individuals per 10k state pop
homelessness["indiv_per_10k"] = 10000 * homelessness["individuals"] / homelessness["state_pop"] 

# Subset rows for indiv_per_10k greater than 20
high_homelessness = homelessness[homelessness["indiv_per_10k"] > 20]

# Sort high_homelessness by descending indiv_per_10k
high_homelessness_srt = high_homelessness.sort_values("indiv_per_10k", ascending=False) 

# From high_homelessness_srt, select the state and indiv_per_10k cols
result = high_homelessness_srt[["state","indiv_per_10k"]]

# See the result
print(result)

                   state  indiv_per_10k
8   District of Columbia      53.738381
11                Hawaii      29.079406
4             California      27.623825
37                Oregon      26.636307
28                Nevada      23.314189
47            Washington      21.829195
32              New York      20.392363


Washington, D.C. has the highest number of homeless individuals - almost 54 per ten thousand people. This is almost double the number of the next-highest state, Hawaii. If you combine new column addition, row subsetting, sorting, and column selection, you can answer lots of questions like this.  
  
### 2. Aggregating DataFrames  
  
**Summary statistics**  
TBD

In [ ]:
# https://campus.datacamp.com/courses/data-manipulation-with-pandas/aggregating-dataframes?ex=1